src/Pate/Verification.hs

```diff
@@ -250,377 +288,67 @@ printResult (Left err) = liftIO $ putStr . ppEquivalenceError $ err
-  rel' <- rel e1 e2
-  justCase <- W4.impliesPred sym bothConds rel'
-  W4.andPred sym eqPreds justCase
-liftPartialRel sym _ W4P.Unassigned W4P.Unassigned = return $ W4.truePred sym
-liftPartialRel sym _ W4P.Unassigned (W4P.PE p2 _) = W4.notPred sym p2
-liftPartialRel sym _ (W4P.PE p1 _) W4P.Unassigned = W4.notPred sym p1
-
-validExit :: MT.ExitCase -> BlockEntryKind arch -> Bool
-validExit ecase blkK = case (ecase, blkK) of
-  (MT.ExitCall, BlockEntryInitFunction) -> True
-  (MT.ExitArch, BlockEntryPostArch) -> True
-  (MT.ExitUnknown, BlockEntryJump) -> True
-  _ -> False
-
-allTargets ::
-  (BlockTarget arch, BlockTarget arch) -> [PatchPair arch]
-allTargets (BlockTarget blkO mrblkO, BlockTarget blkP mrblkP) =
-  [PatchPair blkO blkP] ++
-    case (mrblkO, mrblkP) of
-      (Just rblkO, Just rblkP) -> [PatchPair rblkO rblkP]
-      _ -> []
-
--- | True for a pair of original and patched block targets that represent a valid pair of
--- jumps
-compatibleTargets ::
-  BlockTarget arch ->
-  BlockTarget arch ->
-  Bool
-compatibleTargets blkt1 blkt2 =
-  concreteBlockEntry (targetCall blkt1) == concreteBlockEntry (targetCall blkt2) &&
-  case (targetReturn blkt1, targetReturn blkt2) of
-    (Just blk1, Just blk2) -> concreteBlockEntry blk1 == concreteBlockEntry blk2
-    (Nothing, Nothing) -> True
-    _ -> False
-
--- | Mark a PatchPair as having unsuccessfully attempted verification
-markPairFailed ::
-  PatchPair arch ->
-  EquivM sym arch ()
-markPairFailed pPair = modify $
-  \(EquivState open closed failed) -> EquivState (S.delete pPair open) closed (S.insert pPair failed)
-
 
--- | Mark a PatchPair as having completed verification
-markPairVerified ::
+-- | Verify that the given pre and post condition are satified by the
+-- given block pair. Throws an exception if there is an inequality result, otherwise
+-- it returns normally.
+checkEquivalence ::
   PatchPair arch ->
+  -- | precondition
+  StatePredSpec sym arch ->
+  -- | postcondition
+  StatePredSpec sym arch ->
   EquivM sym arch ()
-markPairVerified pPair = modify $
-  \(EquivState open closed failed) -> EquivState (S.delete pPair open) (S.insert pPair closed) failed
-
-
-addOpenPairs ::
-  Set (PatchPair arch) ->
-  EquivM sym arch ()
-addOpenPairs pPairs = modify $ \(EquivState open closed failed) ->
+checkEquivalence pPair precondSpec postcondSpec = withSym $ \sym -> do
+  withValid @() $ liftIO $ W4B.startCaching sym
+  eqRel <- asks envBaseEquiv
+  stackRegion <- asks envStackRegion
+
+  -- first try proving equivalence by assuming that exact equality
+  -- is the only condition we are propagating backwards, so we
+  -- don't do any work to try to intelligently narrow this down
+
+  -- TODO: it's unclear how fine-grained to make this, or if it's even
+  -- a good idea, but it makes the tests finish in about 1/3 the time
+  result <- manifestError $
+    local (\env -> env { envPrecondProp = PropagateExactEquality }) $
+    provePostcondition pPair postcondSpec
```

Comment:
It would be interesting to collect statistics on how many blocks are exactly equal

Reply:
I agree, but currently this is only at the granularity of functions that were explicitly given (i.e. the 'main' function). So this test is just checking if that top-level entry point can be proven equivalent without any intelligent precondition propagation, and if anything fails anywhere it gives up and starts again from the start.